Fix grid editors placing

thesiv opened the pull request:
Fix the redundant old code in wxGrid::ShowCellEditControl which moves grid
editors but we don't want it. So remove it and also remove workarounds in
editors SetSize functions.

thesiv committed:
Fix grid editors placing

Fix the redundant old code in wxGrid::ShowCellEditControl which moves grid
editors but we don't want it. So remove it and also remove workarounds in
editors SetSize functions.

vadz commented on src/generic/grid.cpp:
```diff
@@ -6975,16 +6975,7 @@ void wxGrid::ShowCellEditControl()
-            if (rect.y > 0)
-                rect.y--;
-#else
+#ifdef __WXQT__
```
I thought we decided to deflate the rectangle by 1 under all platforms to make it fit inside the grid lines?

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -441,38 +441,17 @@ void wxGridCellTextEditor::SetSize(const wxRect& rectOrig)
     // Make the edit control large enough to allow for internal margins
     //
-    // TODO: remove this if the text ctrl sizing is improved esp. for unix
+    // TODO: remove this if the text ctrl sizing is improved
```
I also hoped we could get rid of this TODO... Why is it still necessary and why do we still need platform-specific code here?

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1266,14 +1245,26 @@ void wxGridCellBoolEditor::SetSize(const wxRect& r)
 #if defined(__WXGTK__) || defined (__WXMOTIF__)
+#if defined(__WXGTK3__)
+    cellRect.x += 6;
+    cellRect.width -= 12;
```
This should be written as `cellRect.Deflate(6, 0)` but, more importantly, where does this `6` come from? Is it the same one as in `wxSizerFlags::GetDefaultBorder()`?

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1266,14 +1245,26 @@ void wxGridCellBoolEditor::SetSize(const wxRect& r)
+#if defined(__WXGTK3__)
+    cellRect.x += 6;
+    cellRect.width -= 12;
+#else
```
This seems to be a confusing way to write `#elif defined(__WXMOTIF__)`.

In any case, let's just drop the Motif tests, nobody cares about it anyhow and a wrong offset in a grid editor is the least of its problems.

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1266,14 +1245,26 @@ void wxGridCellBoolEditor::SetSize(const wxRect& r)
-    size.x += 1;
-    size.y -= 2;
+    cellRect.x += 2;
+    cellRect.width -= 2;
```
This would also be more clear to write as a combination of `Deflate(1, 0)` and `Offset(1, 0)`.

But mostly it would be great to get rid of it, instead of tweaking it.

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1266,14 +1245,26 @@ void wxGridCellBoolEditor::SetSize(const wxRect& r)
+    cellRect.width -= 2;
+#elif defined(__WXOSX__)
+    cellRect.x += 3;
+    cellRect.width -= 6;
```
Also `Deflate(3, 0)` and also why `3`?

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1266,14 +1245,26 @@ void wxGridCellBoolEditor::SetSize(const wxRect& r)
 
     // position it in the centre of the rectangle (TODO: support alignment?)
 
+    wxRect cellRect = r;
```
I'd rather keep the old name to avoid the changes in the diff below related to just renaming. If you'd like, let's rename it as a separate commit later, but for now this just makes things less clear.

vadz reviewed (commented):
The new version is slightly better but it's disappointing that we still need all these preprocessor checks. Why do we need them and can we get rid of them entirely, perhaps by adding a new function to retrieve whatever offset we need?

thesiv committed:
fixup! Fix grid editors placing

thesiv commented:
Fixed only checkbox part.

thesiv committed:
fixup! Fix grid editors placing

thesiv committed:
Improve wxRendererNative::GetCheckBoxSize

Add ability to get the size of the checkbox mark without any margins by
passing wxCONTROL_CELL flag.

thesiv commented:
Updated. It's really not so easy to fix platform specific offsets in the text renderer...

vadz commented on src/osx/carbon/renderer.cpp:
```diff
@@ -491,7 +491,7 @@ wxRendererMac::DrawCheckBox(wxWindow *win,
 }
 
-wxSize wxRendererMac::GetCheckBoxSize(wxWindow* win)
+wxSize wxRendererMac::GetCheckBoxSize(wxWindow* win, int UXUNUSED(flags))
```
```suggestion
wxSize wxRendererMac::GetCheckBoxSize(wxWindow* win, int WXUNUSED(flags))
```

vadz commented on include/wx/generic/private/grid.h:
```diff
@@ -1005,5 +1005,12 @@ class WXDLLIMPEXP_ADV wxGridTypeRegistry
+// Returns the rect of the check box in a cell with the given alignmens
+// and the size.
+// The function is used by wxGridCellBoolEditor and wxGridCellBoolRenderer.
+wxRect GetGridCheckBoxRect(const wxSize& checkBoxSize,
```
```suggestion
wxRect wxGetGridCheckBoxRect(const wxSize& checkBoxSize,
```

Even if it's a private function, it's better to still prefix it with `wx` if only for clarity. Another alternative would be to use `wxImpl` or `wxPrivate` namespace, but it's probably not worth it for a single function.

vadz commented on interface/wx/renderer.h:
```diff
@@ -573,8 +573,12 @@ class wxRendererNative
+
+        @param flags The only acceptable flag is @c wxCONTROL_CELL which mean
+            that the size of the only checkbox mark will be returned without
+            any margins.
```
```suggestion
            any margins. This parameter is only available in wxWidgets 3.1.4 or later.
```

vadz commented on src/generic/grid.cpp:
```diff
@@ -10332,4 +10327,44 @@ wxGridCellEditor* wxGridTypeRegistry::GetEditor(int index)
+    wxRect checkBoxRect;
+    checkBoxRect.SetY(cellRect.y + cellRect.height / 2 - checkBoxSize.y / 2);
+
+     wxCoord minSize = wxMin(cellRect.width, cellRect.height);
```
```suggestion
    wxCoord minSize = wxMin(cellRect.width, cellRect.height);
```
Indentation seems to be wrong.

vadz commented on src/generic/grid.cpp:
```diff
@@ -10332,4 +10327,44 @@ wxGridCellEditor* wxGridTypeRegistry::GetEditor(int index)
+     wxCoord minSize = wxMin(cellRect.width, cellRect.height);
+    if ( checkBoxRect.GetWidth() >= minSize || checkBoxRect.GetHeight() >= minSize )
+    {
+        // let the checkbox mark be even smaller then the min size
```
```suggestion
        // let the checkbox mark be even smaller than the min size
```

vadz commented on src/generic/grid.cpp:
```diff
@@ -10332,4 +10327,44 @@ wxGridCellEditor* wxGridTypeRegistry::GetEditor(int index)
+    {
+        // let the checkbox mark be even smaller then the min size
+        // to leave some space between cell edges and the checkbox
+        const int newSize = minSize - 2;
```
What if it's 0 or negative? I think we need to check for this here.

vadz commented on interface/wx/renderer.h:
```diff
@@ -573,8 +573,12 @@ class wxRendererNative
         @param win A valid, i.e. non-null, window pointer which is used to get
             the theme defining the checkbox size under some platforms.
+
+        @param flags The only acceptable flag is @c wxCONTROL_CELL which mean
```
```suggestion
        @param flags The only acceptable flag is @c wxCONTROL_CELL which means
```

vadz commented on interface/wx/renderer.h:
```diff
@@ -573,8 +573,12 @@ class wxRendererNative
             the theme defining the checkbox size under some platforms.
+
+        @param flags The only acceptable flag is @c wxCONTROL_CELL which mean
+            that the size of the only checkbox mark will be returned without
```
```suggestion
            that just the size of the checkbox mark itself will be returned, without
```

vadz commented on src/generic/grideditors.cpp:
```diff
@@ -1237,73 +1216,24 @@ void wxGridCellBoolEditor::Create(wxWindow* parent,
-#ifdef __WXMSW__
-        x += 2;
-#endif
+    wxRect checkBoxRect =
```
```suggestion
    const wxRect checkBoxRect =
```

vadz reviewed (commented):
Thanks, almost good, just a few typos/small fixes here and there (including one which breaks wxOSX build...).

thesiv committed:
fixup! Improve wxRendererNative::GetCheckBoxSize

thesiv committed:
fixup! Fix grid editors placing

thesiv commented:
Fixed.

vadz commented on src/gtk/renderer.cpp:
```diff
@@ -568,26 +570,40 @@ wxRendererGTK::GetCheckBoxSize(wxWindow* win)
+    if ( !addMargins )
+    {
+        gtk_widget_style_get(wxGTKPrivate::GetCheckButtonWidget(),
+                            "indicator_size", &indicator_size);
```
I'll fix this myself, but, just for the record: this is a bad bug (probably an immediate crash) as the list of arguments to this vararg function must be `NULL`-terminated.

vadz commented on src/generic/gridctrl.cpp:
```diff
@@ -16,8 +16,7 @@
 
-#include "wx/generic/gridctrl.h"
-#include "wx/generic/grideditors.h"
+#include "wx/grid.h"
```
Also FYI: I'm not applying this hunk because I don't think it's really needed -- but please let me know if I'm missing something here.

vadz commented on src/generic/gridctrl.cpp:
```diff
@@ -28,9 +27,11 @@
 #endif // WX_PRECOMP
 
-#include "wx/tokenzr.h"
+#include "wx/headerctrl.h"
```
And this one really belongs in `wx/generic/private/grid.h` and not here, which doesn't use anything from that header.

thesiv commented:
Replaced with https://github.com/wxWidgets/wxWidgets/pull/1651.
